convert metric plots to plotly

lboeman opened the pull request:
<!-- Thank you for your contribution! The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items. Feel free to remove checklist items that are not relevant to your change. -->

  - [ ] Closes #xxxx .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [ ] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [ ] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

<!--
Brief description of the problem and proposed solution (if not already fully described in the issue linked to above): -->
Bokeh's interactivity and loading are becoming an issue for larger reports. This converts all metrics plots to plotly. The motivations for this switch are as follows:

- Plotly has a mature and well-documented javascript library. This means that we don't have to rely entirely on the product of some python function call to alter plots. While this may be possible in Bokeh, documentation does not exist or is insufficient. 
- Plotly has a declarative json schema ([https://help.plot.ly/json-chart-schema/](https://help.plot.ly/json-chart-schema/)) which makes storage of these plots convenient. The json string can be read in by either the python or javascript library to reproduce plots. This also saves us from having to store `<script>` elements.

The current code emulates the existing Bokeh plots closely, but there are a couple of things that need to be resolved before this is merged. 

- [x] CSV downloads need to be re-implemented. Each plot now contains it's own data, and we've lost the ability to access the ColumnDataSource on the front end, perhaps this is just accomplished by parshing and formatting the metrics on the report.
- [x] Adapt the parsing of the plotly JSON such that it works for both downloads and rendering in browser, while still achieving the initial intention of lazy-loading the plots after page load.

wholmgren commented:
> This also saves us from having to store <script> elements.

and this presents a security risk to end users that are sharing reports with each other, right?

>  CSV downloads need to be re-implemented. Each plot now contains it's own data, and we've lost the ability to access the ColumnDataSource on the front end, perhaps this is just accomplished by parshing and formatting the metrics on the report.

Should we make an API end point for this?

> Adapt the parsing of the plotly JSON such that it works for both downloads and rendering in browser, while still achieving the initial intention of lazy-loading the plots after page load.

have you done any experiments to determine if this is actually going to make a difference in page loading?

lboeman commented:
> 
> and this presents a security risk to end users that are sharing reports with each other, right?
>
Yes, this makes it more foolproof to ensure we're inserting a malicious script into a template.
 
> 
> Should we make an API end point for this?
> 

If it would be useful beyond this single button, then sure we could add one. Otherwise following the js route seems simpler.
 
> 
> have you done any experiments to determine if this is actually going to make a difference in page loading?
I haven't yet, but the intention here would be to first load the entire document, and then begin inserting metric plots into the correct divs. So we may not end up with a speedup in a completely rendered document, but shifting any long running rendering to the end of the line should improve things.

wholmgren commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1230,6 +1231,14 @@ def __post_init__(self):
+def __check_plot_spec__(plot_spec):
+    """Ensure that the provided plot specification is valid JSON"""
+    try:
+        json.loads(plot_spec)
```
maybe https://python-jsonschema.readthedocs.io/en/stable/ instead?

lboeman replied:
That's an interesting implementation of a 'JSON schema' parser. It doesn't seem to enforce the very opinionated specification proposed by (JSON Schema)[https://json-schema.org/] . 

From tinkering with this briefly, it looks like we'd still have to load the json string into a python dict and then call `validate(instance=json_dict, schema={"type":"object"})` to use this effectively. It does look like it'll give us extra safety though, so that slight performance hit may be worth it?

wholmgren commented on solarforecastarbiter/reports/figures.py:
```diff
@@ -313,7 +336,7 @@ def scatter(timeseries_value_cds, timeseries_meta_cds, units):
 
 
-def construct_metrics_cds(metrics, rename=None):
+def construct_metrics_dataframe(metrics, rename=None):
```
Instead of deleting the existing code, we could create `figures/plotly.py` and `figures/bokeh.py`. Maybe not really needed but multiple plotting "backends" could be useful at some point.

lboeman replied:
I think that's a really good idea. We could add logic to select the processing function based on the sfa reports version. Might require some minor tweaking to the bokeh handling, but I think that'd be a good idea.

lboeman commented:
Added the `reports.figures` module and within it the `bokeh_figures.py` and `plotly_figures.py`. I've adjusted the reports code to only use `plotly_figures.py` and leave the bokeh code there for adventurous core users. Do the changes there seem reasonable or am I off mark on that?

I've also added a script to insert the plotly plots when the standalone html file is generated. Otherwise on the dashboard, the metric_plots variable contains the json to create each plot and the appropriate divs for insertion.

wholmgren commented on docs/source/api.rst:
```diff
@@ -522,15 +522,31 @@ Functions to generate report figures
+   reports.figures.plotly_figures.timeseries_plots
+
+
+Retired functions for generating Bokeh metric plots.
```
```suggestion
Functions for generating Bokeh metric plots.
```

wholmgren replied:
still using time series

wholmgren commented on solarforecastarbiter/reports/figures/bokeh_datamodel.py:
```diff
@@ -0,0 +1,55 @@
+"""File containing necessary datamodel objects for creating bokeh metric
+plots.
```
```suggestion
"""File containing datamodel objects for creating bokeh metric plots.
```

wholmgren commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1230,6 +1233,15 @@ def __post_init__(self):
+
+
 @dataclass(frozen=True)
 class ReportFigure(BaseModel):
```
`PlotlyReportFigure`? Doesn't seem like a library-agnostic class.

lboeman replied:
I think I misunderstood earlier, are you opposed to having an empty `ReportFigure` parent class that gets sub classed by BokehReportFigure and PlotlyReportFigure? I think I mistook your opposition to one class with optional fields as being opposed to that as well.

wholmgren replied:
> are you opposed to having an empty ReportFigure parent class that gets sub classed by BokehReportFigure and PlotlyReportFigure? 

I am not opposed to that. It sounds reasonable. @alorenzo175 should comment before I waste more of your time.

alorenzo175 replied:
sub-classes are ok, but we should make sure they include a ``type`` key so that it's easier to load into ``BokehReportFigure`` or ``PlotlyReportFigure`` when needed

wholmgren commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1250,29 +1262,28 @@ class ReportFigure(BaseModel):
+
 
 @dataclass(frozen=True)
 class RawReportPlots(BaseModel):
```
`PlotlyRawReportPlots`? (as above)

wholmgren commented on solarforecastarbiter/reports/figures/bokeh_figures.py:
```diff
@@ -1,5 +1,12 @@
+Bokeh.
+
+This code is currently unreachable from the rest of the Solar Forecast Arbiter
+Core library. It may be used in place of the plotly_figures to generate bokeh
```
isn't the time series still used?

wholmgren commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -0,0 +1,686 @@
+from bokeh.models import ColumnDataSource, Legend, CDSView, BooleanFilter
+from bokeh.models.ranges import Range1d
+from bokeh.plotting import figure
+from bokeh import palettes
```
no bokeh stuff in the plotly module

wholmgren commented:
>  CSV downloads need to be re-implemented. Each plot now contains it's own data, and we've lost the ability to access the ColumnDataSource on the front end, perhaps this is just accomplished by parshing and formatting the metrics on the report.
Adapt the parsing of the plotly JSON such that it works for both downloads and rendering in browser, while still achieving the initial intention of lazy-loading the plots after page load.

@lboeman seems that the 2nd is now done but not the first?

lboeman commented:
@wholmgren Yes, and I found a bug in the plotting that I'm trying to rectify. You said that we don't necessarily need to display all months, or all days of the week for reports spanning only a few months or days correct?
For csv downloads I think I recall that you wanted to format the csv the same way that we are displaying the metrics tables in reports?

wholmgren commented:
> You said that we don't necessarily need to display all months, or all days of the week for reports spanning only a few months or days correct?

Hours plots should always have 0-24.

Day of week should always have 7 days.

I'm ok with other plots only showing periods for which there is actually data.

lboeman commented:
So it turns out to be a pain to get plotly to display all of the tick labels you supply for a 'categorical' variable. As of this last push, hour of day displays as expected due to the numeric labels. However, the day of the week plots end up like this for, e.g. missing data on saturday
![Screenshot from 2020-03-13 14-09-14](https://user-images.githubusercontent.com/21206164/76659735-670f9280-6534-11ea-9548-0ffc2fb6b25a.png)
There are just missing ticks for the data that should be there.

lboeman commented:
Fixed the weekly plots so that they display a full week regardless of available data. as below:
![Screenshot from 2020-03-16 11-38-24](https://user-images.githubusercontent.com/21206164/76789902-0d4cd980-677b-11ea-8ef7-58e0264439b0.png)

CSV downloads have been implemented by dumping the metrics dataframe as json, and loading it into the html.

wholmgren commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -296,10 +297,19 @@ def bar_subdivisions(df, category, metric):
+            # week of data.
+            y_values = [plot_data[plot_data['index'] == day]['value'].iloc[0]
+                        if not plot_data[plot_data['index'] == day].empty
+                        else 0 for day in x_ticks]
```
can we use `nan` instead of 0?

wholmgren replied:
will the same problem occur with `hourly` if we exclude nighttime values?

lboeman replied:
The hourly plots work correctly when dropping nighttime values. My test report excludes them. For some reason, plotly respects the x_range of an axis when provided as a series of numerical values, but not for things it decides are categorical.

alorenzo175 commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1230,10 +1233,72 @@ def __post_init__(self):
+            return BokehReportFigure.from_dict(dict_, raise_on_extra)
+        else:
+            raise NotImplementedError(
+                f'Do not know how to process {dict_} into a ReportFigure.')
```
should probably not include the dict_ in the message, I imagine it could get big

alorenzo175 replied:
probably also good to remove it from ``BaseFilter``

alorenzo175 commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1253,26 +1318,41 @@ class ReportFigure(BaseModel):
-        plots.
     figures: tuple of :py:class:`solarforecastarbiter.datamodel.ReportFigure`
+    plotly_version: str
+        The plotly version used when generating metrics plots.
```
``bokeh_version``is still a parameter?

lboeman replied:
I've set this up such that it supports `bokeh_version` and `script` or just plotly version. I deleted the `bokeh_datamodel.py` that I had left in. It seemed like we wanted to support continuing to use Bokeh for users who interact directly with the core library?
The reports submodules that process and template the report as well as the actual template files have been altered to only work with the plotly stuff, but if a user was so inclined to create their own Bokeh based report outside of the SFA platform they would have the tools to do so with some slight modifications.

alorenzo175 commented on solarforecastarbiter/reports/figures/bokeh_datamodel.py:
```diff
@@ -0,0 +1,55 @@
+"""File containing datamodel objects for creating bokeh metric
```
Why the separate module? We won't be able to just delete the file when we want to remove bokeh, and it seems confusing to me

alorenzo175 commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -0,0 +1,432 @@
+
+
+@contextmanager
+def _make_webdriver():
```
remove, only for bokeh

alorenzo175 commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -0,0 +1,432 @@
+    return value_cds.data['pair_index'] == pair_index
+
+
+def _extract_metadata_from_cds(metadata_cds, hash_, hash_key):
```
necessary?

alorenzo175 commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1230,10 +1233,72 @@ def __post_init__(self):
+        A descriptive name for the figure.
+    spec: str
+        JSON string representation of the plotly plot.
+    svg: str
```
what would we run into if we wanted this to default to none or empty string and have a pdf field?

lboeman replied:
I don't think it would be much of an issue, we'd just have to add the pdf creation to the RawReport computation, and add templating logic. If there's some gotcha we haven't considered, I don't think it can be too much of an issue since we aren't actively utilizing the svgs for anything. We'd just have to recompute reports so their figures conform to the datamodel.

alorenzo175 replied:
Ok, let's add a pdf parameter that is optional and make an ``output_pdf`` function for the plotly figures

alorenzo175 replied:
or actually, ok to just make an issue and we can do it later

alorenzo175 commented on solarforecastarbiter/reports/figures/tests/test_plotly_figures.py:
```diff
@@ -0,0 +1,316 @@
+            len(report_params.metrics) * len(report_params.object_pairs))
+    )
+
+    # this could maybe use value variance, but asserting the cds process
```
should correct these references to cds here and below

alorenzo175 commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -0,0 +1,432 @@
+
+def output_svg(fig):
+    """
+    Generates an SVG from the Bokeh or Plotly figure. Errors in the
```
only works for plotly. Add note that it requires orca and maybe link to plotly docs for install instructions

alorenzo175 commented on solarforecastarbiter/reports/figures/tests/test_plotly_figures.py:
```diff
@@ -0,0 +1,316 @@
+
+
+@pytest.fixture()
+def no_stray_phantomjs():  # pragma: no cover
```
nix

lboeman commented:
I've opened an issue for adding pdf field to the datamodel.
I believe I've hit all of your collective comments and this is ready for merge. I'll have my full attention back on work on Monday if we'd like to wait for me to be more available to put out any PR related fires. The https://github.com/SolarArbiter/solarforecastarbiter_dashboard/pull/208 PR should display reports adequately, and reports will need to be recomputed to work with this PR.

alorenzo175 commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1230,10 +1233,72 @@ def __post_init__(self):
+    spec: str
+    svg: str
+    figure_type: str
+    figure_class: str = 'plotly'
```
Move this last and possibly make it a ``field(init=False)`` so users can't set it

alorenzo175 commented on solarforecastarbiter/reports/templates/head.html:
```diff
@@ -13,3 +13,8 @@
+{% if plotly_version %}
+{# plotly js and the python plotly library do not have matching versions #}
+{% include "insert_plots.html" %}
+<script src="https://cdn.plot.ly/plotly-1.52.3.min.js"></script>
```
is there any reason to record ``plotly_version`` in the report then?

lboeman replied:
It seems like it might be useful to debug any issues that arise from an older report and newer core library. But maybe just recomputing is the answer there.

alorenzo175 commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -0,0 +1,375 @@
+        svg = fig.to_image(format='svg').decode('utf-8')
+    except Exception:
+        logger.error('Could not generate SVG for figure %s',
+                     getattr(fig, 'name', 'unnamed'))
```
I guess plotly figures don't have name? can you replace name with something that identifies the figure?

alorenzo175 commented on solarforecastarbiter/reports/figures/plotly_figures.py:
```diff
@@ -0,0 +1,375 @@
+        for mvalue in metric_result.values:
+            new = {
+                'name': metric_result.name,
+                'abbrev': f(metric_result.name),
```
any idea why the abbreviate column is the same as name when I run the test report from the cli?

lboeman replied:
I had neglected to pass the abbreviate function when generating the `metrics_json` template var in template.py.

lboeman commented:
@alorenzo175 Think I've hit all your comments. I'm not really opposed to removing `plotly_version` from the report figures, but I think it might be helpful in debugging down the road if we want to do our own version matching between plotly's js library and python.

alorenzo175 commented on solarforecastarbiter/datamodel.py:
```diff
@@ -1255,10 +1257,9 @@ def from_dict(model, input_dict, raise_on_extra=False):
         if model != ReportFigure:
             return super().from_dict(dict_, raise_on_extra)
-        figure_class = dict_.get('figure_class')
-        if figure_class == 'plotly':
```
eee, maybe making init false was a bad idea. I prefer using the figure class to determine this, so probably best to remove that, but keep the parameter at the end

lboeman commented:
> do we need to wait for the dashboard?

Dashboard PR https://github.com/SolarArbiter/solarforecastarbiter_dashboard/pull/208 should work with this PR. Its tests should pass after this is merged. Any existing reports will need to be recomputed. I can take on the API issue to add a recompute endpoint so it'll be easier to roll this out into prod later if you're not already working on it.

alorenzo175 commented:
>  if you're not already working on it.

 I'm not
